### PR TITLE
Run compiler service locally, compile in debug mode if DEBUG is set

### PR DESCRIPTION
**Pull request by jacksonrnewhouse**
This lets users run the compiler service with `
```
BUILD_DIR=$PWD/build_dir OUTPUT_DIR=$PWD/output_dir cargo run arroyo-compiler-service start
```

and then the controller with 
```
REMOTE_COMPILER_ENDPOINT=http://localhost:9000 cargo run  --bin arroyo-controller
```

By default it will build in debug mode, and build times should be under 5 seconds.

### CODE REVIEW

File: arroyo-rpc/build.rs [hunk 1]
```diff
@@ -10,7 +10,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
         .file_descriptor_set_path(out_dir.join("api_descriptor.bin"))
         .type_attribute(".", "#[derive(serde::Serialize, serde::Deserialize)]")
         .type_attribute(".", "#[serde(rename_all = \"camelCase\")]")
-        .compile(&["proto/api.proto"], &["api"])
+        .compile(&["proto/api.proto"], &["proto/"])
```
Comment (mwylde) [inline arroyo-rpc/build.rs:13]: what is this change doing?

**Review by mwylde — COMMENTED** [review]
Couple of questions:

1. Can we consolidate the build_dir introduced here with the existing build_base in docker/? This is adding another thing we need to update when the dependencies of the worker change
2. Have you looked at the performance cost of running pipelines in debug mode? For tokio-heavy code it can be really substantial.

I would also add a big warning somewhere that things are running in debug mode. There's a meme in the rust community at this point that users will come in to the r/rust or stackoverflow asking "why is rust so slow?" and the answer is always that it's running in debug. Since we're changing the default, I would make sure that users are aware of that.

File: .gitignore [hunk 2]
```diff
@@ -2,3 +2,4 @@ target/
 .idea
 .env
 .vscode/
+output_dir/
```
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply .gitignore:5]: Is that an acceptable thing to do? I kinda assumed you weren't supposed to manually do stuff with target/.

File: arroyo-compiler-service/src/main.rs [hunk 3]
```diff
@@ -39,11 +39,13 @@ pub async fn main() {
     let _guard = arroyo_server_common::init_logging("compiler-service");
 
     let build_dir = std::env::var("BUILD_DIR").expect("BUILD_DIR is not set");
+    let is_prod = std::env::var("PROD_PROFILE").is_ok();
```
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply arroyo-compiler-service/src/main.rs:42]: I tried that, but it triggered the PROD variable logging in arroyo-server-common, which wants a log directory.

File: arroyo-compiler-service/src/main.rs [hunk 4]
```diff
@@ -167,9 +170,34 @@ pub struct CompileService {
...
+            .current_dir(&self.build_dir)
+            .arg("build")
+            .arg("--release")
+            .output().await.unwrap()
+        } else {
+            Command::new("cargo")
+            .current_dir(&self.build_dir)
+            .env("RUSTFLAGS", "-A warnings")
+            .arg("build")
+            .arg("--verbose")
```
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply arroyo-compiler-service/src/main.rs:190]: This gives you insight into what it actually needed to rebuild, which is useful for debugging build times. It was very helpful in putting this together, but maybe we don't need it anymore.

File: build_dir/pipeline/src/.gitignore [hunk 5]
```diff
@@ -0,0 +1 @@
+main.rs
```
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply build_dir/pipeline/src/.gitignore:1]: Currently we need the src directories to be present or compilation breaks. Putting a .gitignore is the standard way to commit an empty directory, but I'm down to move it up.

File: arroyo-compiler-service/src/main.rs [hunk 6]
```diff
@@ -167,9 +170,34 @@ pub struct CompileService {
...
+        if self.prod {
+            Command::new("cargo")
+            .current_dir(&self.build_dir)
+            .arg("build")
+            .arg("--release")
+            .output().await.unwrap()
+        } else {
+            Command::new("cargo")
+            .current_dir(&self.build_dir)
+            .env("RUSTFLAGS", "-A warnings")
```
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply arroyo-compiler-service/src/main.rs:188]: This was to get the compiled code warnings away in verbose mode. Realized the right thing to do is just allow them in the file we generate.

File: arroyo-rpc/build.rs (hunk 1 above)
Comment (jacksonrnewhouse) [reply arroyo-rpc/build.rs:13]: I was seeing errors like
       Dirty arroyo-rpc v0.1.0 (/Users/jacksonnewhouse/code/arroyo/arroyo-rpc): the file `/Users/jacksonnewhouse/code/arroyo/arroyo-rpc/api` is missing

Which was causing everything that depends on arroyo-rpc to rebuild on every run. Looking at the code for compile_protos, it should be the directory that the proto definitions are in.

File: arroyo-compiler-service/src/main.rs [hunk 7]
```diff
@@ -167,9 +170,34 @@ pub struct CompileService {
     last_used: Arc<AtomicU64>,
     object_store: Arc<Box<dyn ObjectStore>>,
     base_path: String,
+    prod: bool,
 }
 
 impl CompileService {
+
+    async fn get_command(&self) -> Output {
```
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply arroyo-compiler-service/src/main.rs:178]: Yeah, initially was building the command, not running it.

**Comment by jacksonrnewhouse** [comment]
@mwylde, I think this is ready.

**Pull request merged**